internal/provider/waypoint/resource_waypoint_action.go

```diff
@@ -0,0 +1,607 @@
+					Key:   key,
+					Value: value.ValueString(),
+				})
+			}
+		}
+		if !plan.Request.Custom.URL.IsUnknown() && !plan.Request.Custom.URL.IsNull() {
+			modelBody.ActionConfig.Request.Custom.URL = plan.Request.Custom.URL.ValueString()
+		}
+		if !plan.Request.Custom.Body.IsUnknown() && !plan.Request.Custom.Body.IsNull() {
+			modelBody.ActionConfig.Request.Custom.Body = plan.Request.Custom.Body.ValueString()
+
+		}
+	}
+
+	params := &waypoint_service.WaypointServiceCreateActionConfigParams{
+		NamespaceID: ns.ID,
+		Body:        modelBody,
+	}
+
+	aCfg, err := r.client.Waypoint.WaypointServiceCreateActionConfig(params, nil)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating Action", err.Error())
+		return
+	}
+
+	var aCfgModel *waypoint_models.HashicorpCloudWaypointActionConfig
+	if aCfg.Payload != nil {
+		aCfgModel = aCfg.Payload.ActionConfig
+	}
+	if aCfgModel == nil {
+		resp.Diagnostics.AddError("Unknown error creating Action", "Empty Action returned")
+		return
+	}
+
+	if aCfgModel.ID != "" {
+		plan.ID = types.StringValue(aCfgModel.ID)
+	}
+	if aCfgModel.Name != "" {
+		plan.Name = types.StringValue(aCfgModel.Name)
+	}
+	if aCfgModel.Description != "" {
+		plan.Description = types.StringValue(aCfgModel.Description)
+	}
+
+	plan.ProjectID = types.StringValue(projectID)
+	plan.OrgID = types.StringValue(orgID)
+
+	plan.Request = &actionRequest{}
+
+	if aCfgModel.Request.Custom != nil {
+		diags = readCustomAction(ctx, plan, aCfgModel)
+		if diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+	}
+
+	// Write logs using the tflog package
+	// Documentation: https://terraform.io/plugin/log
+	tflog.Trace(ctx, "Created Action resource")
+
+	// Save data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *ActionResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data *ActionResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectID := r.client.Config.ProjectID
+	if !data.ProjectID.IsUnknown() {
+		projectID = data.ProjectID.ValueString()
+	}
+
+	orgID := r.client.Config.OrganizationID
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: orgID,
+		ProjectID:      projectID,
+	}
+
+	client := r.client
+
+	actionCfg, err := clients.GetAction(ctx, client, loc, data.ID.ValueString(), data.Name.ValueString())
+	if err != nil {
+		if clients.IsResponseCodeNotFound(err) {
+			tflog.Info(ctx, "Action not found for organization, removing from state.")
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError("Error reading Action", err.Error())
+		return
+	}
+
+	if actionCfg.ID != "" {
+		data.ID = types.StringValue(actionCfg.ID)
+	}
+	if actionCfg.Name != "" {
+		data.Name = types.StringValue(actionCfg.Name)
+	}
+	if actionCfg.Description != "" {
```

Comment:
same note on empty here.. if it gets edited remotely, the local state will not pick up on the change